drivers/temperature/ltc2983/ltc2983.c

```diff
@@ -0,0 +1,644 @@
+			descriptor->num_channels++;
+	}
+	if (!descriptor->num_channels) // should be at least one channel
+		goto spi_err;
+
+	ret = no_os_gpio_get(&descriptor->gpio_rstn, &init_param->gpio_rstn);
+	if (ret)
+		goto spi_err;
+	ret = no_os_gpio_direction_output(descriptor->gpio_rstn,
+					  NO_OS_GPIO_LOW);
+	if (ret)
+		goto gpio_err;
+
+	/* bring the device out of reset */
+	no_os_udelay(1200);
+	ret = no_os_gpio_set_value(descriptor->gpio_rstn, NO_OS_GPIO_HIGH);
+	if (ret)
+		goto gpio_err;
+
+	ret = ltc2983_setup(descriptor);
+	if (ret)
+		return ret;
+
+	*device = descriptor;
+	return 0;
+
+gpio_err:
+	no_os_gpio_remove(descriptor->gpio_rstn);
+spi_err:
+	no_os_spi_remove(descriptor->comm_desc);
+free_err:
+	no_os_free(descriptor);
+	return ret;
+}
+
+/**
+ * @brief Remove resources allocated by the init function
+ * @param device - LTC2983 descriptor
+ * @return 0 in case of success, errno errors otherwise
+ */
+int ltc2983_remove(struct ltc2983_desc *device)
+{
+	int ret;
+	if (!device)
+		return -ENODEV;
+
+	ret = no_os_gpio_remove(device->gpio_rstn);
+	if (ret)
+		return -EINVAL;
+
+	ret = no_os_spi_remove(device->comm_desc);
+	if (ret)
+		return -EINVAL;
+
+	no_os_free(device);
+
+	return 0;
+}
+
+/**
+ * @brief Read raw register value
+ * @param device - LTC2983 descriptor
+ * @param reg_addr - register address
+ * @param val - register value
+ * @return 0 in case of success, negative error code otherwise
+ */
+int ltc2983_reg_read(struct ltc2983_desc *device, uint16_t reg_addr,
+		     uint8_t *val)
+{
+	int ret;
+	uint8_t raw_array[] = {LTC2983_SPI_READ_BYTE, reg_addr >> 8,
+			       reg_addr & 0xFF, 0x0
+			      };
+
+	ret = no_os_spi_write_and_read(device->comm_desc, raw_array,
+				       NO_OS_ARRAY_SIZE(raw_array));
+	if (ret)
+		return ret;
+	*val = raw_array[3];
+
+	return 0;
+}
+
+/**
+ * @brief Write raw register value
+ * @param device - LTC2983 descriptor
+ * @param reg_addr - register address
+ * @param val - register value
+ * @return 0 in case of success, negative error code otherwise
+ */
+int ltc2983_reg_write(struct ltc2983_desc *device, uint16_t reg_addr,
+		      uint8_t val)
+{
+	int ret;
+	uint8_t raw_array[] = {LTC2983_SPI_WRITE_BYTE, reg_addr >> 8,
+			       reg_addr & 0xFF, val
+			      };
+
+	ret = no_os_spi_write_and_read(device->comm_desc, raw_array,
+				       NO_OS_ARRAY_SIZE(raw_array));
+	if (ret)
+		return ret;
+
+	return 0;
+}
+
+/**
+ * @brief Update register value
+ * @param device - LTC2983 descriptor
+ * @param reg_addr - register address
+ * @param mask - Mask for specific register bits to be updated
+ * @param val - register value (requires prior bit shifting)
+ * @return 0 in case of success, negative error code otherwise
+ */
+int ltc2983_reg_update_bits(struct ltc2983_desc *device, uint16_t reg_addr,
+			    uint8_t mask, uint8_t val)
+{
+	int ret;
+	uint8_t data;
+
+	ret = ltc2983_reg_read(device, reg_addr, &data);
+	if (ret)
+		return ret;
+
+	data = (data & ~mask) | (val & mask);
```

Comment:
For other drivers, we usually assume the `val` is actually the value of the masked field, without any shifting. I think we should do the same here, since it's more straight forward for the function caller.